pipeline-worker/src/pipeline-steps/utils.ts

```
@@ -0,0 +1,153 @@
+  }
+  const cost =
+    0.001 *
+    (tokIn * COST_BY_MODEL[modelName].in_per_1K +
+      tokOut * COST_BY_MODEL[modelName].out_per_1K);
+  return success(cost);
+}
+
+/**
+ * Create a logger with report context for better debugging across concurrent reports
+ *
+ * @param module - Module name for logging context
+ * @param userId - Optional user ID for context
+ * @param reportId - Optional report ID for context
+ * @returns Pino logger with context
+ */
+export function getReportLogger(
+  module: string,
+  userId?: string,
+  reportId?: string,
+): Logger {
+  const context: Record<string, string> = { module };
+  if (userId) {
+    context.userId = userId;
+  }
+  if (reportId) {
+    context.reportId = reportId;
+  }
+  return logger.child(context);
+}
+
+/**
+ * Process a batch of items concurrently with concurrency limiting
+ *
+ * This utility processes items in batches with a maximum concurrency limit,
+ * useful for rate-limited API calls or memory-constrained operations.
+ *
+ * @param items - Array of items to process
+ * @param processor - Async function to process each item
+ * @param concurrency - Maximum number of concurrent operations
+ * @returns Promise that resolves to array of results in the same order as input
+ *
+ * @example
+ * const results = await processBatchConcurrently(
+ *   comments,
+ *   async (comment) => await processComment(comment),
+ *   6  // Process up to 6 comments concurrently
+ * );
+ */
+export async function processBatchConcurrently<T, R>(
+  items: T[],
+  processor: (item: T) => Promise<R>,
+  concurrency: number,
+): Promise<R[]> {
+  const results: R[] = new Array(items.length);
+  const executing: Set<Promise<void>> = new Set();
+
+  for (let i = 0; i < items.length; i++) {
+    const index = i;
+    const promise = processor(items[i])
+      .then((result) => {
+        results[index] = result;
+      })
+      .finally(() => {
+        executing.delete(promise);
```

If processor did fail, we probably do want to log that though. I think this would cause silent failures.